flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecMatch.java

```diff
@@ -0,0 +1,440 @@
+                        new CodeGeneratorContext(config),
+                        planner.createRelBuilder(),
+                        false, // nullableInput
+                        JavaScalaConversionUtil.toScala(cepPatternAndNames.f1),
+                        JavaScalaConversionUtil.toScala(Optional.empty()),
+                        CodeGenUtils.DEFAULT_COLLECTOR_TERM());
+        generator.bindInput(
+                inputRowType,
+                CodeGenUtils.DEFAULT_INPUT1_TERM(),
+                JavaScalaConversionUtil.toScala(Optional.empty()));
+        final PatternProcessFunctionRunner patternProcessFunction =
+                generator.generateOneRowPerMatchExpression(
+                        (RowType) getOutputType(), partitionKeys, matchSpec.getMeasures());
+        final CepOperator<RowData, RowData, RowData> operator =
+                new CepOperator<>(
+                        inputSerializer,
+                        isProcTime(inputRowType),
+                        nfaFactory,
+                        eventComparator,
+                        cepPattern.getAfterMatchSkipStrategy(),
+                        patternProcessFunction,
+                        null);
+        final OneInputTransformation<RowData, RowData> transform =
+                ExecNodeUtil.createOneInputTransformation(
+                        timestampedInputTransform,
+                        createTransformationMeta(MATCH_TRANSFORMATION, config),
+                        operator,
+                        InternalTypeInfo.of(getOutputType()),
+                        timestampedInputTransform.getParallelism());
+        final RowDataKeySelector selector =
+                KeySelectorUtil.getRowDataSelector(partitionKeys, inputTypeInfo);
+        transform.setStateKeySelector(selector);
+        transform.setStateKeyType(selector.getProducedType());
+
+        if (inputsContainSingleton()) {
+            transform.setParallelism(1);
+            transform.setMaxParallelism(1);
+        }
+        return transform;
+    }
+
+    public void checkOrderKeys(RowType inputRowType) {}
+
+    private EventComparator<RowData> createEventComparator(
+            ExecNodeConfig config, RowType inputRowType) {
+        SortSpec orderKeys = matchSpec.getOrderKeys();
+        if (orderKeys.getFieldIndices().length > 1) {
+            GeneratedRecordComparator rowComparator =
+                    ComparatorCodeGenerator.gen(
+                            config, "RowDataComparator", inputRowType, orderKeys);
+            return new RowDataEventComparator(rowComparator);
+        } else {
+            return null;
+        }
+    }
+
+    public Transformation<RowData> translateOrder(
+            Transformation<RowData> inputTransform, RowType inputRowType, ReadableConfig config) {
+        return inputTransform;
+    }
+
+    @VisibleForTesting
+    public static Tuple2<Pattern<RowData, RowData>, List<String>> translatePattern(
```

Comment:
ditto

Reply:
@godfreyhe, `translatePattern` is called by `org.apache.flink.table.planner.match.PatternTranslatorTestBase`, hence this uses `public`.